satellite/audit/worker.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/storj"
+)
+
+// Worker contains information for populating audit queue and processing audits.
+type Worker struct {
+	log     *zap.Logger
+	queue   *Queue
+	Loop    sync2.Cycle
+	Limiter sync2.Limiter
+}
+
+// NewWorker instantiates Worker.
+func NewWorker(log *zap.Logger, queue *Queue, config Config) (*Worker, error) {
+	return &Worker{
+		log: log,
+
+		queue:   queue,
+		Loop:    *sync2.NewCycle(config.QueueInterval),
+		Limiter: *sync2.NewLimiter(config.WorkerConcurrency),
+	}, nil
+}
+
+// Run runs audit service 2.0.
+func (worker *Worker) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	worker.log.Info("starting")
+
+	// wait for all audits to run
+	defer worker.Limiter.Wait()
+
+	return worker.Loop.Run(ctx, func(ctx context.Context) error {
+		err := worker.process(ctx)
```

Comment:
@egonelbre , if we queue all the work in the queue to the limiter, then the queue gets updated by the chore, then we call `worker.process(ctx)` again here after the queue interval, I think the following could happen:
* The queue is updated by the chore to have 1000 unique items in it
* 1000 items of work are added to the limiter, and the queue is now empty
* The queue is updated by the chore again, and it has 1000 items again. However, the limiter has only managed to process 500 of the originally queued items. 
* The worker loop runs again, and adds the 1000 items from the queue onto the limiter. Now we have 1500 items to process, but we don't really care about the first 500 anymore because we got a new queue.
I have some ideas in mind, but I don't know if they are the best way to solve the issue.

Reply:
The limiter ensures that only up to `N` workers can run at a time. If the limit is reached it waits. So if there are 1000 unique items, it will wait after 3 have been added.

Reply:
Oh, I understand better now. For some reason I was under the impression that `worker.Limiter.Go` would still return if `N` workers were in progress, and that `Go` would still process that work at a later point. After taking another look at the limiter code, I see that I was mistaken.